### PR TITLE
Add CLI API

### PR DESCRIPTION
**Pull request by pankgeorg**
CLI API:
(Web API coming soon!)
```
usage: <PROGRAM> [--host HOST] [--port PORT] [--dry-run]
                 [--notebook_paths [NOTEBOOK_PATHS...]]
                 [--skip-run-notebooks [SKIP-RUN-NOTEBOOKS...]]
                 [--export] [--export-only]
                 [--export-exclude [EXPORT-EXCLUDE...]] [--enable-ui]
                 [--separate-pluto-state] [--binder-url BINDER-URL]
                 [--slider-server-url SLIDER-SERVER-URL]
                 [--cache-dir CACHE-DIR] [--output-dir OUTPUT-DIR]
                 [-c CONFIG] [--sample-toml SAMPLE-TOML]
                 [--enable-api] [--secret SECRET] [--version] [-h]
                 [start_dir]

Pluto Slide Server. A boutique pluto-notebook runner that gives life
to your static exports.
Can respond to @bind macros updates but also just create static
exports that look cool.

positional arguments:
  start_dir             The starting directory (default: ".")

optional arguments:
  --host HOST           host:port (default: "127.0.0.1")
  --port PORT           port (type: Int64, default: 2345)
  --dry-run             Stop the the slider server after running once.
                        Useful in conjuction with --export.
  --notebook_paths [NOTEBOOK_PATHS...]
                        The paths in which to look for notebooks
                        (type: Vector{String})
  --skip-run-notebooks [SKIP-RUN-NOTEBOOKS...]
                        list of notebook files to skip from running.
                        (type: Vector{String})
  --export              also export static files
  --export-only         Same as --export --dry-run
  --export-exclude [EXPORT-EXCLUDE...]
                        Export option. list of notebook files to skip
                        from exporting (type: Vector{String})
  --enable-ui           Export option. Don't hide all buttons and
                        toolbars. Without this flag, they will be
                        hidden by default to make the page look like
                        an article
  --separate-pluto-state
                        Export option. If set this flag splits the
                        export to two files: a .plutostate and an
                        .html. Default is to have state 'baked in' the
                        file'
  --binder-url BINDER-URL
                        Export option. This will set the binder URL
                        that will be used as a backend. MIT Math uses:
                                                https://mybinder.org/v2/gh/mitmath/18S191/e2dec90
                        (default: "")
  --slider-server-url SLIDER-SERVER-URL
                        Export option. The URL of the slider server
                        which can respond to requests of @bind
                        elements
  --cache-dir CACHE-DIR
                        if provided, use this directory to read and
                        write cached notebook states. Caches will be
                        indexed by notebook hash, but you need to take
                        care to invalidate the cache when Pluto or
                        this export script updates. Useful in
                        combination with
                        https://github.com/actions/cache.
  --output-dir OUTPUT-DIR
                        folder to write generated HTML files to (will
                        create directories to preserve the input
                        folder structure). Leave at the default to
                        generate each HTML file in the same folder as
                        the notebook file. (default: ".")
  -c, --config, --configuration_file CONFIG
                        Use this option to provide a configuration
                        TOML
  --sample-toml SAMPLE-TOML
                        Print a sample TOML configuration file
  --enable-api          Allow API requests
  --secret SECRET       Web API secret
  --version             show version information and exit
  -h, --help            show this help message and exit
```

### CODE REVIEW

**Comment by pankgeorg** [comment]
To test the cli, run this at the project where you `dev` Pluto & PlutoBindServer: 
`julia --project=. -e "using PlutoSliderServer; cli()" -- --help` (note the `--` before `--help`)

**Comment by fonsp** [comment]
Wow!!

**Comment by fonsp** [comment]
We could use https://github.com/Roger-luo/Comonicon.jl if it's useful

**Comment by fonsp** [comment]
Did you also see https://github.com/fonsp/Pluto.jl/pull/1029 ? Let's talk about combining that PR with SliderServer, maybe instead of a CRUD api?

I'm thinking: the PR uses the enter-exit-update idea from d3 to update only the necessary parts of the notebook. We can also do that on the folder level:

the SliderServer launches on a directory, and we could rescan the directory, boot up notebooks that were not there before, shut down notebooks that got deleted, and update changed notebook files using https://github.com/fonsp/Pluto.jl/pull/1029

File: src/WebAPI.jl [hunk 1]
```diff
@@ -1,12 +1,47 @@
...
-import PlutoSliderServer: FinishedNotebookSession, RunningNotebookSession, QueuedNotebookSession, myhash, MoreAnalysis
+import PlutoSliderServer: FinishedNotebookSession, RunningNotebookSession, QueuedNotebookSession, myhash, MoreAnalysis, find_notebook_files_recursive
 
 # curl -X DELETE http://127.0.0.1:2345/notebook/uut+KtbdWht451eFX7gBhnqfHRSarm0KK6NyontJkQE=/
 # curl -X GET http://127.0.0.1:2345/notebook/uut+KtbdWht451eFX7gBhnqfHRSarm0KK6NyontJkQE=/
 # curl -X POST --data-binary "@./notebook2.jl.assets/notebook2.jl" http://127.0.0.1:2345/notebook/start/
+# curl -X GET http://127.0.0.1:2345/git_pull_and_rescan/?github_url=https://github.com/pankgeorg/spam
 
+function path_hash(path)
+    myhash(open(f -> read(f, String), path))
```
Comment (fonsp) [inline src/WebAPI.jl:12]: myhash(read(path))

**Comment by pankgeorg** [comment]
image
!!!!

File: src/WebAPI.jl [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            # get the state object
+            original_state = Pluto.notebook_to_js(notebook)
+            # shut down the notebook (later)
+            try_tocache(settings.Export.cache_dir, hash, original_state)
+            if keep_running
+                bond_connections = MoreAnalysis.bound_variable_connections_graph(notebook)
+                @info "Bond connections" showall(collect(bond_connections))
+                session = RunningNotebookSession(;
+                    path=path,
+                    hash=path_hash(path),
```
Comment (pankgeorg) [inline src/WebAPI.jl:44]: bug!

File: src/WebAPI.jl [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+                run(`rm -rf "$folder"`)
+                # Clone without history
+                # Fetch/Pull if you have latest
+                # Also have some cleanup around!
+                run(`git clone "$toclone"`)
+            else 
+                return HTTP.Response(501, "Can't pull")
+            end
+            start_dir = "$this_folder/$folder"
+            settings, pluto_options = get_configuration("$this_folder/$folder/pluto-deployment-environment/PlutoDeployment.toml")
```
Comment (pankgeorg) [inline src/WebAPI.jl:252]: if that is different, restart 
write a file (.env~!!!~~~) that the process will find after boot
and run `shutdown()`

**Pull request closed without merging**